### ISSUE TITLE
Incorrect computation of BCELoss on cuda when tensor sizes are different

### ISSUE THREAD

**Issue by devhello145** [pytorch/pytorch#40023]
## 🐛 Bug

Computation of BCELoss on Cuda is incorrect, when sizes of tensors are differ.
There is a UserWarning, but it's only _warning_.
On CPU device BCELoss is OK.

## To Reproduce

1. Run code sample
2. See result in CPU mode (same shape and different shape results are equal)
3. See result in GPU mode (same shape and different shape results are differ)
```
import numpy as np
import torch
import torch.nn.functional as F


def compute_grad(a, gt, device):
    a_t = torch.tensor(a, requires_grad=True, device=device)
    # a_t.retain_grad()

    gt_t = torch.tensor(gt).to(device)

    criterion = torch.nn.BCELoss()

    # shape is equal
    loss = criterion(a_t, gt_t)
    loss.backward()
    sum_grad1 = torch.sum(a_t.grad).item()

    # size is differ
    a_t.grad.data.zero_()
    loss = criterion(a_t.unsqueeze(1), gt_t)  # add dimension
    loss.backward()
    sum_grad2 = torch.sum(a_t.grad).item()
    return sum_grad1, sum_grad2


if __name__ == "__main__":

    np.random.seed(42)

    N = 1000
    bs = 100
    a = np.random.rand(bs, N).astype(np.float64)

    true_index = np.random.randint(N, size=N // 2).reshape(1, N // 2)
    gt = np.zeros((bs, N), dtype=np.float64)
    gt[:, true_index] = 1

    cpu_result = compute_grad(a, gt, device=torch.device("cpu"))
    gpu_result = compute_grad(a, gt, device=torch.device("cuda:0"))

    print("CPU: same shape: {} \t diff shape: {}".format(*cpu_result))
    print("GPU: same shape: {} \t diff shape: {}".format(*gpu_result))

    # CPU: same shape: 1.9732641300852216      diff shape: 1.9732641300852216
    # GPU: same shape: 1.973264130085222       diff shape: 197.3264130085222

```
## Expected behavior

I expect to get the same result in gpu mode as in cpu or raise an error

## Environment
```
Collecting environment information...
PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.17.3

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration: GPU 0: GeForce RTX 2060 SUPER
Nvidia driver version: 440.59
cuDNN version: /usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip3] numpy==1.18.4
[pip3] pytorch-warmup==0.0.4
[pip3] torch==1.5.0
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @ngimel

**Comment by glaringlee** [comment]
mark high priority since cpu and gpu give different behavior.

**Comment by ngimel** [comment]
cc @kurtamohler #31365. I believe current behavior is expected, but perhaps we should really disable it going forward or at the very least make sure that cpu and gpu give the same results.

**Comment by kurtamohler** [comment]
It was a while ago when I worked on this, but if I remember correctly and if I'm interpreting my comment here (https://github.com/pytorch/pytorch/pull/31365#issuecomment-575839479) correctly, this difference existed in the old TH implementation had this issue as well. We decided to keep the incorrect behavior and just print a warning because it would be BC breaking to either change it or to throw an error. I think the plan was to add an error in the next pytorch release, and then fix the behavior in the release after that.
